### PR TITLE
Fixes #15409 - Separated puppet facts from core

### PR DESCRIPTION
**Pull request by ShimShtein**
Refactored facts import mechanism so, that puppet facts could be separated.
1. Moved SmartProxy reference from `#import_host` to `#import_facts` - the source is relevant for fact retrieval process, not to host creation.
2. Moved `#import_host` to `Host::Base`: it's base service and it is not limited to managed hosts.
3. Redesigned `#populate_fields_from_facts` to be responsible for filling host's fields from a predefined fact parser.
4. Refactored `FactParser` and `FactImporter` registries to expose a default option and removed references to puppet from there.

### CODE REVIEW

**Comment by ShimShtein** [comment]
Split a problematic method to two and used `:send` to call `include` outside of the class definition.

**Comment by ShimShtein** [comment]
Rubocop issue fixed

**Comment by ShimShtein** [comment]
The error seems unrelated (in upgrade scenario it couldn't retrieve `gem-empty` gem).
[test]

**Comment by xprazak2** [comment]
Could you rebase pls?

**Comment by ShimShtein** [comment]
Rebased and reviewed. Should be OK now.

**Comment by xprazak2** [comment]
I am getting:

```
2017-05-28T08:04:08 db015a7c [app] [I] Started POST "/api/hosts/facts" for ::1 at 2017-05-28 08:04:08 +0000
2017-05-28T08:04:08 db015a7c [app] [I] Processing by Api::V2::HostsController#facts as JSON
2017-05-28T08:04:08 db015a7c [app] [I]   Parameters: {"facts"=>"[FILTERED]", "name"=>"shore-patrol.example.com", "certname"=>"shore-patrol.example.com", "apiv"=>"v2", :host=>{"certname"=>"shore-patrol.example.com", "name"=>"shore-patrol.example.com"}}
2017-05-28T08:04:08 db015a7c [app] [D] Importer Katello::RhsmFactImporter does not implement authorized_smart_proxy_features.
2017-05-28T08:04:08 db015a7c [app] [D] Importer ForemanAnsible::StructuredFactImporter does not implement authorized_smart_proxy_features.
D, [2017-05-28T08:04:08.679301 #22113] DEBUG -- : String does not start with the prefix 'encrypted-', so Setting::Auth restrict_registered_smart_proxies was not decrypted
2017-05-28T08:04:08 db015a7c [app] [D] String does not start with the prefix 'encrypted-', so Setting::Auth restrict_registered_smart_proxies was not decrypted
D, [2017-05-28T08:04:08.685512 #22113] DEBUG -- : String does not start with the prefix 'encrypted-', so Setting::Auth require_ssl_smart_proxies was not decrypted
2017-05-28T08:04:08 db015a7c [app] [D] String does not start with the prefix 'encrypted-', so Setting::Auth require_ssl_smart_proxies was not decrypted
D, [2017-05-28T08:04:08.691940 #22113] DEBUG -- : String does not start with the prefix 'encrypted-', so Setting::Auth ssl_client_cert_env was not decrypted
2017-05-28T08:04:08 db015a7c [app] [D] String does not start with the prefix 'encrypted-', so Setting::Auth ssl_client_cert_env was not decrypted
D, [2017-05-28T08:04:08.695200 #22113] DEBUG -- : String does not start with the prefix 'encrypted-', so Setting::Auth ssl_client_dn_env was not decrypted
2017-05-28T08:04:08 db015a7c [app] [D] String does not start with the prefix 'encrypted-', so Setting::Auth ssl_client_dn_env was not decrypted
D, [2017-05-28T08:04:08.720552 #22113] DEBUG -- : String does not start with the prefix 'encrypted-', so Setting::Auth trusted_puppetmaster_hosts was not decrypted
2017-05-28T08:04:08 db015a7c [app] [D] String does not start with the prefix 'encrypted-', so Setting::Auth trusted_puppetmaster_hosts was not decrypted
2017-05-28T08:04:08 db015a7c [app] [D] Verifying request from ["localhost", "centos-devel.example.com", "centos-devel"] against ["centos-devel.example.com", "centos-devel.example.com"]
2017-05-28T08:04:08 db015a7c [app] [I] Current user: foreman_api_admin (administrator)
2017-05-28T08:04:08 db015a7c [app] [D] Setting current user thread-local variable to foreman_api_admin
2017-05-28T08:04:08 db015a7c [app] [W] Action failed
 | ArgumentError: wrong number of arguments (2 for 3..4)
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/foreman-tasks-0.8.6/app/models/foreman_tasks/concerns/host_action_subject.rb:18:in `import_host'
 | /home/vagrant/projects/foreman/app/controllers/api/v2/hosts_controller.rb:271:in `facts'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/actionpack-4.2.8/lib/action_controller/metal/implicit_render.rb:4:in `send_action'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/actionpack-4.2.8/lib/abstract_controller/base.rb:198:in `process_action'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/actionpack-4.2.8/lib/action_controller/metal/rendering.rb:10:in `process_action'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/actionpack-4.2.8/lib/abstract_controller/callbacks.rb:20:in `block in process_action'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:117:in `call'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:117:in `call'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:555:in `block (2 levels) in compile'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:505:in `call'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:505:in `call'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:498:in `block (2 levels) in around'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:313:in `call'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:313:in `block (2 levels) in halting'
 | /home/vagrant/projects/foreman/app/controllers/api/v2/base_controller.rb:152:in `disable_json_root'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:432:in `block in make_lambda'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:312:in `call'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:312:in `block in halting'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:497:in `call'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:497:in `block in around'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:505:in `call'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:505:in `call'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:498:in `block (2 levels) in around'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:313:in `call'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:313:in `block (2 levels) in halting'
 | /home/vagrant/projects/foreman/app/controllers/concerns/application_shared.rb:15:in `set_timezone'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:432:in `block in make_lambda'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:312:in `call'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:312:in `block in halting'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:497:in `call'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:497:in `block in around'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:505:in `call'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:505:in `call'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:498:in `block (2 levels) in around'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:313:in `call'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:313:in `block (2 levels) in halting'
 | /home/vagrant/projects/foreman/app/models/concerns/foreman/thread_session.rb:32:in `clear_thread'
```

**Comment by ShimShtein** [comment]
@xprazak2, It looks like an unnecessary monkey patch in foreman-tasks. I have already created a PR there to remove this monkey patch. In the meantime, could you check this PR again with `FOREMAN_TASKS_MONKEYS` set to false?

**Comment by xprazak2** [comment]
I tested with `FOREMAN_TASKS_MONKEYS=true bundle exec rails s`, I got a different error:

```
2017-05-30T10:16:43 395b1a8d [foreman-tasks/action] [E] wrong number of arguments (4 for 1..3) (ArgumentError)
 | /home/vagrant/projects/foreman/app/models/host/base.rb:106:in `import_host'
 | /home/vagrant/projects/foreman-tasks/app/lib/actions/foreman/host/import_facts.rb:11:in `plan'
```

**Comment by ShimShtein** [comment]
Fixed foreman-tasks in https://github.com/theforeman/foreman-tasks/pull/251. @xprazak2, can you apply it and then retest?

File: app/models/concerns/facets/managed_host_extensions.rb
Comment (xprazak2) [inline app/models/concerns/facets/managed_host_extensions.rb]: I think there should be `source_proxy` instead of `proxy`. When I try to change it I get:

```
2017-06-06T08:34:58 3cf3a422 [app] [W] Action failed
 | NoMethodError: undefined method `populate_fields_from_facts' for #<Katello::Host::ContentFacet:0x0000000e1187b8>
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activemodel-4.2.8/lib/active_model/attribute_methods.rb:433:in `method_missing'
 | /home/vagrant/projects/foreman/app/models/concerns/facets/managed_host_extensions.rb:94:in `block in populate_facet_fields'
 | /home/vagrant/projects/foreman/app/models/concerns/facets/managed_host_extensions.rb:93:in `each'
 | /home/vagrant/projects/foreman/app/models/concerns/facets/managed_host_extensions.rb:93:in `populate_facet_fields'
 | /home/vagrant/projects/foreman/app/models/host/managed.rb:428:in `populate_fields_from_facts'
 | /home/vagrant/projects/foreman/app/models/concerns/puppet_host_extensions.rb:9:in `populate_fields_from_facts_with_puppet'
 | /home/vagrant/projects/foreman/app/models/host/base.rb:156:in `parse_facts'
 | /home/vagrant/projects/foreman/app/models/host/base.rb:145:in `import_facts'
 | /home/vagrant/projects/foreman/app/models/host/managed.rb:292:in `import_facts'
 | /home/vagrant/projects/foreman/app/controllers/api/v2/hosts_controller.rb:272:in `facts'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/actionpack-4.2.8/lib/action_controller/metal/implicit_render.rb:4:in `send_action'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/actionpack-4.2.8/lib/abstract_controller/base.rb:198:in `process_action'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/actionpack-4.2.8/lib/action_controller/metal/rendering.rb:10:in `process_action'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/actionpack-4.2.8/lib/abstract_controller/callbacks.rb:20:in `block in process_action'
```

**Comment by xprazak2** [comment]
We are not quite there yet...
```
2017-06-06T08:22:20 2f85716c [app] [W] Action failed
 | NameError: undefined local variable or method `proxy' for #<Host::Managed:0x007f54aa766de8>
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activemodel-4.2.8/lib/active_model/attribute_methods.rb:433:in `method_missing'
 | /home/vagrant/projects/foreman/app/models/concerns/facets/managed_host_extensions.rb:94:in `block in populate_facet_fields'
 | /home/vagrant/projects/foreman/app/models/concerns/facets/managed_host_extensions.rb:93:in `each'
 | /home/vagrant/projects/foreman/app/models/concerns/facets/managed_host_extensions.rb:93:in `populate_facet_fields'
 | /home/vagrant/projects/foreman/app/models/host/managed.rb:428:in `populate_fields_from_facts'
 | /home/vagrant/projects/foreman/app/models/concerns/puppet_host_extensions.rb:9:in `populate_fields_from_facts_with_puppet'
 | /home/vagrant/projects/foreman/app/models/host/base.rb:156:in `parse_facts'
 | /home/vagrant/projects/foreman/app/models/host/base.rb:145:in `import_facts'
 | /home/vagrant/projects/foreman/app/models/host/managed.rb:292:in `import_facts'
 | /home/vagrant/projects/foreman/app/controllers/api/v2/hosts_controller.rb:272:in `facts'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/actionpack-4.2.8/lib/action_controller/metal/implicit_render.rb:4:in `send_action'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/actionpack-4.2.8/lib/abstract_controller/base.rb:198:in `process_action'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/actionpack-4.2.8/lib/action_controller/metal/rendering.rb:10:in `process_action'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/actionpack-4.2.8/lib/abstract_controller/callbacks.rb:20:in `block in process_action'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:117:in `call'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:117:in `call'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:555:in `block (2 levels) in compile'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:505:in `call'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:505:in `call'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:498:in `block (2 levels) in around'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:313:in `call'
 | /home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:313:in `block (2 levels) in halting'
 | /home/vagrant/projects/foreman/app/controllers/api/v2/base_controller.rb:152:in `disable_json_root'
```

**Comment by ShimShtein** [comment]
Fixed the bug and added a test.

**Comment by xprazak2** [comment]
Nice, my facts were uploaded and I got no errors.

**Comment by ohadlevy** [comment]
@ShimShtein this needs a rebase again?

**Comment by ShimShtein** [comment]
Rebased again...

**Comment by adamruzicka** [comment]
What is the status here?

**Comment by ShimShtein** [comment]
Waiting for review... :smile:

**Comment by xprazak2** [comment]
I have no additional comments, it was working for me.

**Review by adamruzicka — APPROVED** [review]
Code wise looks good to me, @xprazak2 tested it and didn't have any additional comments. Let's get this in

File: app/models/host/base.rb
Comment (iNecas) [inline app/models/host/base.rb]: Please update the deprecation version

**Comment by ShimShtein** [comment]
Rebased and moved deprecation to 1.19.

**Comment by iNecas** [comment]
[test katello]

**Comment by mmoll** [comment]
@ShimShtein could you please rebase to also have the Rails 5 tests running against this?

**Comment by ShimShtein** [comment]
Rebased to the latest develop.

**Comment by mmoll** [comment]
and it's broken :)

**Comment by ShimShtein** [comment]
Updated deprecation whitelist line numbers.

**Comment by mmoll** [comment]
and it's fixed 👍

File: app/models/concerns/puppet_host_extensions.rb
Comment (mmoll) [inline app/models/concerns/puppet_host_extensions.rb]: It would be nice to have this done in a more Rails 5 friendly way, see GH-4845

**Comment by ShimShtein** [comment]
Changed `alias_method_chain` to `prepend`.

**Comment by mmoll** [comment]
unfortunately this needs another rebase, as the deprecation was fixed now.

**Comment by ShimShtein** [comment]
Rebased again.

**Comment by iNecas** [comment]
[test foreman]

**Comment by iNecas** [comment]
Almost there: I would like to make the https://github.com/theforeman/foreman-tasks/pull/251#issuecomment-330500189 changes compatible with foreman 1.16, as otherwise the tasks master would not be fully compatible with next stable master until quite long time. Ready to go after that is addressed.

**Comment by iNecas** [comment]
One last rebase please before merge, nothing else is blocking us from merging this one

**Comment by ShimShtein** [comment]
Rebased

**Comment by iNecas** [comment]
Thanks @ShimShtein for patience. Only 1 year and something to merge :)

**Comment by iNecas** [comment]
Also kudos to reviewers

**Comment by lzap** [comment]
And discovery is all red again. From red to red. Time to add discovery into core test suite? Just stable ruby version on sqlite3 would be fine I guess...

File: app/models/host/base.rb
Comment (lzap) [inline app/models/host/base.rb:131]: The `set_taxonomies` call needs to be done before `populate_fields_from_facts`. Created https://github.com/theforeman/foreman/pull/6478 to fix this.

**Pull request merged**
